ja/documentation/ruby-from-other-languages/index.md

```diff
@@ -0,0 +1,441 @@
+`Fixnum`に時間を扱うメソッド群を定義しています。
+
+{% highlight ruby %}
+class Fixnum
+  def hours
+    self * 3600 # 1時間あたりの秒数を返します
+  end
+  alias hour hours
+end
+
+# 1月1日の0時0分から14時間後
+Time.mktime(2006, 01, 01) + 14.hours # => Sun Jan 01 14:00:00
+{% endhighlight %}
+
+### 不思議なメソッド名
+
+Rubyでは、メソッド名の最後に疑問符(?)や感嘆符(!)が使われることがあります。
+慣習的に、問い合わせ系のメソッドの最後には疑問符が使われます
+(例: レシーバが空の場合に`true`を返す`Array#empty?`)。
+また、副作用のあるメソッドの最後には感嘆符が使われます
+(例: `self`あるいは引数の内容を書き換えるようなメソッド。`exit!`など)。
+けれど、副作用のあるメソッドすべてがこの慣習に従っているわけでも
+ないことに注意してください。
+`Array#replace`は、与えられた配列の内容で配列の内容を置き換えます。
+この操作は、自身を**変更しない**ようなメソッド名とは意に反する振る舞いをします。
+
+### 特異メソッド
+
+特異メソッドはオブジェクト単位のメソッドです。
+特異メソッドは、定義したオブジェクトからだけ利用できるメソッドになります。
+
+{% highlight ruby %}
+class Car
+  def inspect
+    "Cheap car"
+  end
+end
+
+porsche = Car.new
+porsche.inspect # => Cheap car
+def porsche.inspect
+  "Expensive car"
+end
+
+porsche.inspect # => Expensive car
+
+# 他のオブジェクトには影響しない
+other_car = Car.new
+other_car.inspect # => Cheap car
+{% endhighlight %}
+
+### 「存在しなかった」メソッド
+
+Rubyはメッセージに対応するメソッドを見つけられなかったとしても諦めません。
+その場合は、見つけられなかったメソッド名と引数と共に、
+`method_missing`メソッドを呼び出します。
+`method_missing`メソッドはデフォルトではNameError例外を投げますが、
+アプリケーションに合うように再定義することもできます。
+実際、多くのライブラリがそのようにしています。
+以下がその例です。
+
+{% highlight ruby %}
+# IDは呼び出されたメソッド名です。
+# *構文は「arguments」という名前の配列に
+# 呼び出されたメソッドに渡されたすべての引数を格納します。
+def method_missing(id, *arguments)
+  puts "Method #{id} was called, but not found. It has " +
+       "these arguments: #{arguments.join(", ")}"
+end
+
+__ :a, :b, 10
+# => Method __ was called, but not found. It has these
+# arguments: a, b, 10
+{% endhighlight %}
+
+このコードは単に呼び出されたメソッドの詳細を出力しているだけですが、
+ここには何を書いても良いので、メッセージを受け取って適切な処理を行えます。
+
+### 関数呼び出しではなくメッセージ送信
+
+メソッド呼び出しは実際には他のオブジェクトへの**メッセージ送信**です。
+
+{% highlight ruby %}
+# これは
+1 + 2
+# これと同じで...
+1.+(2)
+# つまりこういうことです。
+1.send "+", 2
+{% endhighlight %}
+
+### ブロックは作りかけのオブジェクト
+
+ブロック(実際にはクロージャ)は標準ライブラリでもすごく使われています。
+ブロックを呼び出すには、`yield`を使うか、引数リストに特別な引数を追加して
+それを`Proc`オブジェクトにします。以下がその例です。
+
+{% highlight ruby %}
+def block(&the_block)
+  # この内側では、the_blockはメソッドに渡されたブロックになります。
+  the_block # ブロックが返ります
+end
+adder = block { |a, b| a + b }
+# adderはここでProcオブジェクトになります
+adder.class # => Proc
+{% endhighlight %}
+
+`Proc.new`にブロックを渡すか、`lambda`メソッドを呼び出すことで、
+メソッド呼び出しの外側でブロックを作成することもできます。
+
+同様に、メソッドもまた作りかけのオブジェクトです。
+
+{% highlight ruby %}
+method(:puts).call "puts is an object!"
+# => putsはオブジェクト!
+{% endhighlight %}
+
+### 演算子は糖衣構文(シンタックスシュガー)
+
+Rubyにおけるほとんどの演算子は糖衣構文です。
+いくつかの優先順位規則にもとづいて、メソッド呼び出しを単に書き換えているだけです。
+たとえば、Fixnumクラスの`+`メソッドを次のようにオーバーライドすることもできます。
+
+{% highlight ruby %}
+class Fixnum
+  # できるけれど、しないほうがいいでしょう
+  def +(other)
+    self - other
+  end
+end
+{% endhighlight %}
+
+C++の`operator+`などは必要ありません。
+
+`[]`や`[]=`といったメソッドを定義すれば、配列のようなスタイルでアクセスすることもできます。
+(+1や-2みたいな)単項の+や-を定義するには、`+@`か`-@`といったメソッドを定義する必要があります。
+けれど、以下の演算子は糖衣構文では**ありません**。
+これらはメソッドではないので、再定義できません。
+
+{% highlight ruby %}
+=, .., ..., !, not, &&, and, ||, or, !=, !~, ::
```

Comment:
`!`, `!=`, and `!~` are overloadable since 1.9.1 https://github.com/ruby/ruby/blob/trunk/doc/NEWS-1.9.1#L227-L228